samples/nrf_desktop/src/hw_interface/buttons.c

```diff
@@ -252,22 +252,26 @@ static void scan_fn(struct k_work *work)
 	/* Emit event for any key state change */
 	static u32_t old_state[COLUMNS];
 	bool any_pressed = false;
+	size_t evt_limit = 0;
 
 	for (size_t i = 0; i < COLUMNS; i++) {
 		for (size_t j = 0; j < ARRAY_SIZE(row); j++) {
 			bool is_pressed = cur_state[i] & BIT(j);
 			bool was_pressed = old_state[i] & BIT(j);
 
-			if (is_pressed != was_pressed) {
+			if ((is_pressed != was_pressed) &&
+			    (evt_limit < CONFIG_DESKTOP_BUTTONS_EVENT_LIMIT)) {
 				struct button_event *event = new_button_event();
 
 				event->key_id = (i << 8) | (j & 0xFF);
 				event->pressed = is_pressed;
 				EVENT_SUBMIT(event);
-			}
 
-			any_pressed = any_pressed || is_pressed;
+				evt_limit++;
+			}
 		}
+
+		any_pressed = any_pressed || (cur_state[i] != 0);
 	}
 
 	memcpy(old_state, cur_state, sizeof(old_state));
```

Comment:
this part is wrong - I forgot to change it
we need to update old_state bit by bit to make sure events will be sent in the subsequent scan

Reply:
fixed